metronome/rocksdb/src/io/iohk/metronome/rocksdb/RocksDBStore.scala

```diff
@@ -0,0 +1,374 @@
+  // isolated from reads via locks, so for example if we
+  // read an ID, then retrieve the record from a different
+  // collection, we can be sure it hasn't been deleted in
+  // between the two operations.
+  private val lockRead    = Sync[F].delay(rwlock.readLock().lock())
+  private val unlockRead  = Sync[F].delay(rwlock.readLock().unlock())
+  private val lockWrite   = Sync[F].delay(rwlock.writeLock().lock())
+  private val unlockWrite = Sync[F].delay(rwlock.writeLock().unlock())
+
+  private def withReadLock[A](fa: F[A]): F[A] =
+    Sync[F].bracket(lockRead)(_ => fa)(_ => unlockRead)
+
+  private def withWriteLock[A](fa: F[A]): F[A] =
+    Sync[F].bracket(lockWrite)(_ => fa)(_ => unlockWrite)
+
+  // In case there's a write operation among the reads and we haven't
+  // taken out a write lock, we can upgrade the read lock for the duration
+  // of the write, then downgrade it back to read when we're done.
+  // See here for the rules up upgrading/downgrading:
+  // https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReentrantReadWriteLock.html
+  private def withLockUpgrade[A](fa: F[A]): F[A] =
+    Sync[F].bracket {
+      unlockRead >> lockWrite
+    }(_ => fa) { _ =>
+      lockRead >> unlockWrite
+    }
+
+  /** Execute the accumulated write operations in a batch. */
+  private val writeBatch: ReaderT[F, BatchEnv, Unit] =
+    ReaderT {
+      case (opts, batch) if batch.hasPut() || batch.hasDelete() =>
+        Sync[F].delay {
+          db.write(opts, batch)
+          batch.clear()
+        }
+      case _ =>
+        ().pure[F]
+    }
+
+  /** Execute one `Get` operation. */
+  private def read[K, V](op: Get[Namespace, K, V]): F[Option[V]] = {
+    for {
+      kbs <- encode(op.key)(op.keyCodec)
+      mvbs <- Sync[F].delay[Option[Array[Byte]]] {
+        Option(db.get(handles(op.namespace), kbs))
+      }
+      mv <- mvbs match {
+        case None =>
+          none.pure[F]
+
+        case Some(bytes) =>
+          decode(bytes)(op.valueCodec).map(_.some)
+      }
+    } yield mv
+  }
+
+  /** Collect writes in a batch, until we either get to the end, or there's a read.
+    * This way writes are atomic, and reads can see their effect.
+    *
+    * In the next version of RocksDB we can use transactions to make reads and writes
+    * run in isolation.
+    */
+  private val batchingCompiler: KVNamespacedOp ~> Batch =
+    new (KVNamespacedOp ~> Batch) {
+      def apply[A](fa: KVNamespacedOp[A]): Batch[A] =
+        fa match {
+          case op @ Put(n, k, v) =>
+            ReaderT { case (_, batch) =>
+              for {
+                kbs <- encode(k)(op.keyCodec)
+                vbs <- encode(v)(op.valueCodec)
+                _ = batch.put(handles(n), kbs, vbs)
+              } yield ()
+            }
+
+          case op @ Get(_, _) =>
+            // Execute any pending deletes and puts before performing the read.
+            writeBatch >> ReaderT.liftF(read(op))
```

Comment:
I wonder if there is a way to restrict such case as it seams a little arbitray at this level execute write when the real operation was read.

Do you know if it is posible to `KVStoreOp` have lets say two sub classes like:
```
sealed trait KVStoreOp[N, A]
sealed train ReadOp[N, A] extends KVStoreOp[N, A]
sealed train WriteOp[N, A] extends KVStoreOp[N, A]
```
and then having operations to inherit from correct subclasses and then compilers to take correct operation sublclasses ?

Reply:
Well it's still a batch program, like a doobie `ConnectionIO[_]` or whatever it's called, that too can perform reads and writes in one transaction. You can imagine that a compiler could just do `beginTransaction()` and do read and writes alone, I just thought this is probably more efficient and would probably be what you want to do most of the time, with the patterns we already have: write to half a dozen different stores.

I thought about what you are suggesting, and I think what's required is a different free monad type:
```
type KVStoreRead[N, A] = Free[({ type L[A] = KVStoreOp.Get[N, A] })#L, A]
```
You'd have to roll similar compilers to turn a `KVStoreRead` to your monad of choice, and then you can execute multiple reads but no writes.

Reply:
I didn't want to duplicate the whole stack, but I'll look at the Inject API to see if we can from the beginning treat them as two different Free monads that can be combined

Reply:
I tried with the `InjectK` stuff but I ended up where I was with more difficult to understand code. 

What I did in the end was to introduce the type above, `KVStoreRead`, and added a `.readonly` property to `KVCollection` which gives you an interface that only does `.get`. There's a `KVStore.Ops.lift` method to turn it into the read-write version. 

So this could be used like this:

```
object BlockHeaderStore extends KVCollection[Namespace, Hash, BlockHeader]
object BlockBodyStore extends KVCollection[Namespace, Hash, BlockHeader]

val query: KVStoreRead[Namespace, Block] = for {
  h <- BlockHeaderStore.readonly.get(blockHash)
  b <- BlockBodyStore.readonly.get(blockHash)
} yield Block(blockHash, h, b)
```